Range selector

cmangin opened the pull request:
This patch adds range selectors for tactics.
You can write "[1, 3-5]: tac." to apply the tactic "tac" to the goals numbered 1, 3, 4 and 5.

gasche commented on printing/ppvernac.ml:
```diff
@@ -967,8 +967,13 @@ module Make
+        in
         let pr_goal_selector = function
           | SelectNth i -> int i ++ str":"
+          | SelectList l -> prlist_with_sep (fun () -> str ", ") pr_range_selector l
```
will this print the `[`, `]` back?

cmangin replied:
It would, yes. I noticed there was a similar problem with the "[id]" and "par" selectors, too, I'll submit another PR for that.

gasche commented on proofs/proofview.ml:
```diff
@@ -380,6 +380,37 @@ let tclFOCUS_gen nosuchgoal i j t =
+    (* Sort the list to get the left-most goal to focus. This goal won't
+       move, and we will then place all the other goals to focus to the
+       right. *)
+    let l = CList.sort compare l in
```
if this is `Pervasives.compare`, I think you should prefix it explicitly -- @ppedrot is rather suspicious of them and he wants to be able to locate them quickly.

cmangin replied:
I remove it altogether, since apparently there is no guarantee that the order of `Pervasives.compare` on pairs is the lexicographic one.

gasche commented on proofs/proofview.ml:
```diff
@@ -380,6 +380,37 @@ let tclFOCUS_gen nosuchgoal i j t =
+    match l with
+      | [] -> tclZERO (NoSuchGoals 0)
+      | (mi, _) :: _ ->
+          let left, sub_right = CList.goto (mi-1) comb in
```
For the non-expert reader, the code would be more readable if `left` was named `rev_left`.

cmangin replied:
I used the same names as in the function [focus_sublist] higher in the file, but I'll add a comment at least to make it more clear that [left, sub_right] is a zipper.

mattam82 commented on proofs/proofview.mli:
```diff
@@ -239,6 +239,16 @@ val set_nosuchgoals_hook: (int -> string option) -> unit
 val tclFOCUS : int -> int -> 'a tactic -> 'a tactic
 
+(** [tclFOCUSLIST li t] applies [t] on the list of focused goals
+    described by [li]. Each element of [li] is a pair [(i, j)] denothing
```
denothing

cmangin replied:
Thanks, fixed it.

mattam82 commented:
LGTM, @aspiwack is it ok with you?

aspiwack commented:
Yes, the implementation is rather economical, and should make sense. Documenting the behaviour would be a good thing, though.

gares commented:
2 simple questions:
- are the [] really needed? (I don't think so, but I want to check if you tried to parse without and failed or not)
- I want this for ; too, like tac;2-4:now auto.  Now that you know the code, do you think it is easy to code? If so, I can try to hack it myself.

gares commented:
Maybe I should be more explicit about why I don't like the [].  If the feature is also supported in other ltac constructs like [|] and ; then the extra brackets are not making things simpler/nicer. Compare tac;2-4:tac and tac;[2-4]:tac or tac;[2-4:tac|..] and tac;[[2-4]:tac|..].  The option with no brackets is not ambiguous and less confusing to me: When you see [ it is always a branching, no lookahead needed.

To be fair, Arnaud told me at the CS that [name] is already there to select goals by name.  So brackets have the merit of being consistent with that (new) feature.  Still, IIRC, ?[name] is a way to name evars corresponding to goals.  If I'm not wrong, then ?[name]: tac seems more consistent (with the sytanx of terms) to me.

coqdev commented:
While we are at it :-). I would like this too:
1,3-4:{
Intros.
Apply too.
}

Le mer. 1 juil. 2015 à 08:33, Enrico <notifications@github.com> a écrit :

> 2 simple questions:
>
>    - are the [] really needed? (I don't think so, but I want to check if
>    you tried to parse without and failed or not)
>    - I want this for ; too, like tac;2-4:now auto. Now that you know the
>    code, do you think it is easy to code? If so, I can try to hack it myself.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coq/coq/pull/74#issuecomment-117458014>.
>

herbelin commented:
About the brackets. I used brackets for idents for parsing purpose, because any tactic can start with an ident. Similarly, under the assumption that we adopt "tac1; 2-3:tac2", and want to write" Ltac f n1 n2 := tac1; n1-n2:tac2", we would have parsing issues by not using a specific token (but do we really want this? my first impression is that this is hackish style).

More generally, I would like to brainstormingly consider the opportunity:

1) for selecting goals by names (typically by constructor name in an induction case), which, with a good naming scheme, can be more robust and intuitive than numbers (at least in induction cases).
2) for bullets to select a goal, and hence to have a syntax for that. I don't have a clear view at what would be optimal here. I was thinking e.g at "+ [S n {IHn}] " to tell that I want to focus on the successor case, naming the predecessor n and the induction hypothesis IHn. However, in an interactive session, this would mean the "[S n]" to visibly change the focus first made by +, which is inelegant.
An alternative would be "+[S n]" (but not elegant w/o space), or "[S n] +" (not elegant as not highlighting the indentation), or simply [S n], considering that [] is an alternative syntax for focusing, but this starts to make things complicated.

gares commented:
To me "tac0;[solve tac1| | | solve tac1 | | ]" is less readable than "tac0;1,4: solve tac1" and in some cases more efficient than "tac0; try solve tac1".  I guess we should ask the Ltac power users (like Adam).  If they don't need that, then I won't push further the idea.

cmangin commented:
Just to answer the initial technical questions:

* Currently, the syntax without the brackets would immediately conflict with the still existing syntax `1: tac`, which is translated to a `SelectNth` instead of a `SelectList`. The reason I kept the `SelectNth`, which is theoretically a particular case of `SelectList`, is performance: every tactic without a goal selector is interpreted with the default goal selector (which can be user-defined). In effect, it means that for every tactic, there is a `SelectNth`, which is not a matter since it's quite fast. I was afraid replacing it completely with the appropriate `SelectList` would be too heavy (but I guess I should have benchmarked this). A solution could be to remove `SelectNth` at the parsing level, but not at the default goal selector level (so that the default goal selector can be `SelectNth`, but anything that the user prepends to a tactic is a `SelectList`). Or maybe performance is not a problem actually :-) .

* I don't think having `tac1; 1,3: tac2` would be too much work, except for possible syntax problems. It would just require to move completely the selectors into the TacExpr part. Currently, the goal selectors are parsed in `parsing/g_vernac.ml4` instead of `parsing/g_ltac.ml4` like the other tacticals. So it could just be a tactical like any other.

JasonGross commented:
> ask the Ltac power users (like Adam)

Do I count?  I tend to be happy with either `foo; try solve [ bar ]` or `foo; [ bar | | bar ]` or `foo; [ .. | bar ]`.  I infrequently have enough side conditions that `[ | ... | ]` is unwieldy, and when I do, I prefer to either throw brute force at them, or pick them out by the shape of the goal, to be robust to side-condition reordering.

gasche commented:
@cmangin yeah it would be easy to have a single parsing rule and keep a `SelectNth` constructor for the fast path -- and indeed I'm sure that you can implement `SelectList` so that there is no observable performance difference for one-element lists.

aspiwack commented:
A few technical remarks.

First: `1:tac` and `[1]:tac` currently do not have quite the same behaviour: in the former, `tac` is interpreted in the environment of the first goals, and in the latter in the global environment. Consider the following:

```coq
Goal True -> True.
Proof.
  intros y.
  Fail [1]:let x := y in idtac x.
  let x := y in idtac x.
```

Not that the behaviour of `[1]:tac` is terribly useful, so I don't think it would be a problem to parse a single goal as `SelectNth` and selectors with several goals as `SelectList` (plus, conceivably, we could have the current `[1]:tac` by writing `1-1:tac`). That's if the parser is ok with it. Coq's parser can be quite fragile.

Now, about having selectors as tacticals. Let me first remark that they are indeed currently (and that's still the case with @cmangin's proposal) implemented as tacticals. They are just not parsed as such (more precisely, they are parsed as command who are all intepreted as calling a tactic on the current proof, the actual tactic is modified by the selector). So in principle, the only potential obstacle for having such tacticals is parsing.

Also note that for reasons not unrelated to my first point, `[1-2]:tac` does not have the same semantics as `[>tac|tac|..]`, the latter being goal local. This can be observed with

```coq
Goal True /\ True.
Proof.
  split.
  [1-2]: let n := numgoals in idtac "Number of goals" n.
    (* Number of goals 2 *)
  all:[> let n := numgoals in idtac "Number of goals" n | let n := numgoals in idtac "Number of goals" n | ..].
   (* Number of goals 1
       Number of goals 1 *)
```

So it may be a useful tactical to have.

maximedenes commented:
test build

coqdev commented:
Stop merci                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        De: Maxime DénèsEnvoyé: lundi 6 juillet 2015 17:42À: coq/coqRépondre à: coq/coqCc: The coqdev team (anonymous account)Objet: Re: [coqdev] [coq] Range selector (#74)test build

—Reply to this email directly or view it on GitHub.

maximedenes commented:
test build

cmangin commented:
After trying to hack the `tac1; 2: tac2` idea, I came to a few conclusions.
* I believe we need to differentiate between a selector for a whole command in proof-edition mode and a selector in front of some tactic, because there is not much sense to write `tac1; par: tac2`, but we want to be able to write `par: tac1; tac2`.
* Following that, the default goal selector (which can currently be `all:` or `1:`) should be a "top-level" selector, because a default selector in front of `tac2` in `tac1; tac2` would not be intuitive (currently, the selector for `tac2` would actually be `all:` (sort of...), not `1:`, anyway).
* Then there is the problem of parsing: how do you differentiate between `1: tac` where `1:` is a top-level selector and where `1:` is a "tactical" selector?

What I am suggesting is to have two different classes of selectors. The top-level selectors, which can be one of `all:`, `par:` or nothing (with nothing being the selector `1:`) ; and the tactical selectors, which can be either an identifier `[id]:`, a list of ranges `n1, n2-n3:` or nothing (with nothing being the selector `all:`). The `SelectNth` constructor would still be kept to address the problem raised by @aspiwack's first example.

I feel like it should be intuitive enough for the user and clean enough for the implementation, but maybe I am missing something. Any thought?

@aspiwack: about your example, I am wondering why this works (it prints `<var>` instead of `y`, but it is accepted):
```coq
Goal True \/ True -> True.
Proof.
  intros [y|y]; let x := y in idtac x. (* <var> *)
```
What should be the expected behaviour there?

Note that `intros [y1|y]; let x := y in idtac x.` also works, but not `intros [y1|y2]; let x := y in idtac x.`.

aspiwack commented:
@cmangin Note that there is an option (`Set Default Goal Selector`) to change the default goal selector (the default default (!) goal selector is `1:`, I once considered suggesting to use `all:` instead but after some experimenting I changed my mind, I've been considering instead a selector `just-one` (non-committal name) which would raise an error if more than one goal is available).

Your surprising examples come from the limitations of Ltac's binding analysis. It's unrelated to the current discussion.

cmangin commented:
A few months later, an update on this.

I implemented the idea of using goal selectors as tacticals (and dropping the brackets), so you can write this proof for instance:
```coq
Goal True /\ (True /\ True).
Proof.
  split; 2: split; exact I.
Qed.
```
or this one:
```coq
Goal exists x, x = 0.
Proof.
  eexists ?[x]; ?[x]: exact True; reflexivity.
Qed.
```
A few comments:

1. Notice that you have to write `?[x]:` instead of just `[x]:` as before. The reason is that now that this construction is a tactical, there would be an ambiguity between `[x]: tac.` and `[tac].` that you cannot solve easily when parsing. (For some reason, I couldn't have `?x` being accepted by the parser, but I don't know why.)
2. Currently, this binds more tightly than the semicolon, which means in the first example above that it is parsed as `split; (2: split); exact I`. I found it to be more natural, but it is open to discussion.
3. I noticed that even if there exists an existential named `?x`, you cannot focus it by its name, unless it was explicitely named by the user at some point, as in the second example above. Is it an intentional restriction?
4. I disallowed the use of the `par: ` selector inside a tactic at runtime, mainly because I was not sure when this should raise an error (but I'm pretty sure that it should be disallowed...right?). What this means is that you can write `Ltac foo := par: idtac.` and this will succeed, even though the use of `foo` will fail everytime.
5. The interaction with bullets when giving negative feedback is not very good: if you have used a bullet, have only one goal, and do `2: idtac`, it will answer `The current bullet - is unfinished.`. This is not a trivial fix, because the functions that implement focusing for bullets and this PR are shared, and this error message is produced by a hook that just checks if there is an open bullet.

About possible extensions:

1. This does not implement the idea of `1,3-4: { ... }` that was suggested. Do we want this? I don't think that it would be too difficult. I have no idea how this would interact with other changes like @gares' proof blocks, though.
2. While we are at it, is the possibility of writing `1, ?[x], 3-4, ?[y]: tac` a good idea?

aspiwack commented:
> About possible extensions:

I think it is advisable to merge this without further delay (I think @ppedrot is best placed to green-light it). The PR had been ok-d a year ago, so unless someone raises objections (in particular on the recent commits), I believe any further work should be in a separate PR.

> Notice that you have to write ?[x]: instead of just [x]: as before

I have a vague memory of a bug report where the `[x]:` syntax created a syntax conflict. So I guess this solves the issue. I believe this syntax was not documented in v8.5, so maybe it is a good idea to backport the syntax change to v8.5 as as well (if this syntax is merged).

I think @herbelin should validate this.

> Currently, this binds more tightly than the semicolon

Sounds reasonable to me

> The interaction with bullets when giving negative feedback is not very good: if you have used a bullet, have only one goal, and do 2: idtac, it will answer The current bullet - is unfinished.. This is not a trivial fix, because the functions that implement focusing for bullets and this PR are shared, and this error message is produced by a hook that just checks if there is an open bullet.

This is not a regression, though. You can observe this behaviour in current Coq. I had devised the code for focusing primitives to be rather generic, but it resulted in poor error messages, especially for bullet. @matafou has done the bullet-specific error messages. The code is a bit _ad hoc_ but the error messages, while not perfect, are really better.

gares commented:
par today does not work in a nested ltac expressions.  I think it would be better to give a parse error rather than failing at runtime.  In any case such failure must be different from a regular tactic failure.

cmangin commented:
I agree with this, but I was not sure of the cleanest way to do it, code-wise. Here is an attempt.
(I'm actually wondering if `all:` should even be allowed, since it does not have any effect inside a ltac expression.)

Matafou commented:
2016-05-31 23:25 GMT+02:00 Arnaud Spiwack <notifications@github.com>:

> This is not a regression, though. You can observe this behaviour in
> current Coq. I had devised the code for focusing primitives to be rather
> generic, but it resulted in poor error messages, especially for bullet.
> @matafou <https://github.com/matafou> has done the bullet-specific error
> messages. The code is a bit *ad hoc* but the error messages, while not
> perfect, are really better.
>
>
Indeed the code is ad hoc. Actually I would like to dive into it once more
to implement other informative messages: printing unfocused goals with
bullets (and curly braces) informations after the goal (or on demand). Like
this:

H1: ...
H2: ...
============================
foo y z

*** Unfocused goals:
   * forall x:nat, bar x
   * forall x y:nat, bar x z }
 + forall x, foo' x }
 + ..
- ...
